src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs

```diff
@@ -179,78 +176,94 @@ public bool TryGetPlatformVersion(string versionString, [NotNullWhen(true)] out
                 }
                 return Version.TryParse(versionString, out version);
             }
-
-            return false;
         }
 
         public string GetVersionString(string framework, Version version)
         {
-            var versionString = string.Empty;
-
-            if (version != null
-                && (version.Major > 0
-                    || version.Minor > 0
-                    || version.Build > 0
-                    || version.Revision > 0))
+            if (version is null || IsZero(version))
             {
-                var versionParts = new Stack<int>(4);
+                return string.Empty;
+            }
+
+            int major = version.Major > 0 ? version.Major : 0;
+            int minor = version.Minor > 0 ? version.Minor : 0;
+            int build = version.Build > 0 ? version.Build : 0;
+            int revision = version.Revision > 0 ? version.Revision : 0;
 
-                versionParts.Push(version.Major > 0 ? version.Major : 0);
-                versionParts.Push(version.Minor > 0 ? version.Minor : 0);
-                versionParts.Push(version.Build > 0 ? version.Build : 0);
-                versionParts.Push(version.Revision > 0 ? version.Revision : 0);
+            // remove all trailing zeros beyond the minor version
+            int partCount = (minor == 0, build == 0, revision == 0) switch
+            {
+                (true, true, true) => 1,
+                (false, true, true) => 2,
+                (_, false, true) => 3,
+                (_, _, false) => 4
+            };
 
+            if (partCount == 1)
+            {
                 // By default require the version to have 2 digits, for legacy frameworks 1 is allowed
                 var minPartCount = SingleDigitVersionFrameworks.Contains(framework) ? 1 : 2;
 
-                // remove all trailing zeros beyond the minor version
-                while ((versionParts.Count > minPartCount
-                       && versionParts.Peek() <= 0))
-                {
-                    versionParts.Pop();
-                }
+                partCount = Math.Max(partCount, minPartCount);
```

Comment:
nit: this can probably be partCount?

Reply:
Fixed. I did a little more refactoring for consistency and readability.

For example, we used the set `SingleDigitVersionFrameworks` to check for three frameworks in one place, but then chained three `Equals` calls in another to check three different frameworks.